nsxt/metadata/metadata.go

```diff
@@ -0,0 +1,256 @@
+	for key, item := range metadata {
+		if item.Metadata.Skip {
+			continue
+		}
+
+		log.Printf("[INFO] inspecting key %s", key)
+		if len(parent) > 0 {
+			log.Printf("[INFO] parent %s key %s", parent, key)
+		}
+		if elem.FieldByName(item.Metadata.SdkFieldName).IsNil() {
+			log.Printf("[INFO] skip key %s with nil value", key)
+			continue
+		}
+		if item.Metadata.SchemaType == "struct" {
+			nestedObj := elem.FieldByName(item.Metadata.SdkFieldName)
+			nestedSchema := make(map[string]interface{})
+			childElem := item.Schema.Elem.(*ExtendedResource)
+			StructToSchema(nestedObj.Elem(), d, childElem.Schema, key, nestedSchema)
+			log.Printf("[INFO] assigning struct %+v to %s", nestedObj, key)
+			var nestedSlice []map[string]interface{}
+			nestedSlice = append(nestedSlice, nestedSchema)
+			if len(parent) > 0 {
+				parentMap[key] = nestedSlice
+			} else {
+				d.Set(key, nestedSlice)
+			}
+		} else if item.Metadata.SchemaType == "list" || item.Metadata.SchemaType == "set" {
+			if _, ok := item.Schema.Elem.(*ExtendedSchema); ok {
+				// List of string, bool, int
+				nestedSlice := elem.FieldByName(item.Metadata.SdkFieldName)
+				log.Printf("[INFO] assigning slice %v to %s", nestedSlice.Interface(), key)
+				if len(parent) > 0 {
+					parentMap[key] = nestedSlice.Interface()
+				} else {
+					d.Set(key, nestedSlice.Interface())
+				}
+			} else if childElem, ok := item.Schema.Elem.(*ExtendedResource); ok {
+				// List of struct
+				sliceElem := elem.FieldByName(item.Metadata.SdkFieldName)
+				var nestedSlice []map[string]interface{}
+				for i := 0; i < sliceElem.Len(); i++ {
+					nestedSchema := make(map[string]interface{})
+					StructToSchema(sliceElem.Index(i), d, childElem.Schema, key, nestedSchema)
+					nestedSlice = append(nestedSlice, nestedSchema)
+					log.Printf("[INFO] appending slice item %+v to %s", nestedSchema, key)
+				}
+				if len(parent) > 0 {
+					parentMap[key] = nestedSlice
+				} else {
+					d.Set(key, nestedSlice)
+				}
+			}
+		} else {
+			if len(parent) > 0 {
+				log.Printf("[INFO] assigning nested value %+v to %s", elem.FieldByName(item.Metadata.SdkFieldName).Interface(), key)
+				parentMap[key] = elem.FieldByName(item.Metadata.SdkFieldName).Interface()
+			} else {
+				log.Printf("[INFO] assigning value %+v to %s", elem.FieldByName(item.Metadata.SdkFieldName).Interface(), key)
+				d.Set(key, elem.FieldByName(item.Metadata.SdkFieldName).Interface())
+			}
+		}
+	}
+}
+
+// SchemaToStruct converts terraform schema to NSX model struct
+// currently supports nested subtype and trivial types
+func SchemaToStruct(elem reflect.Value, d *schema.ResourceData, metadata map[string]*ExtendedSchema, parent string, parentMap map[string]interface{}) {
+	for key, item := range metadata {
+		if item.Metadata.ReadOnly || item.Metadata.Skip {
+			continue
+		}
+		if item.Metadata.IntroducedInVersion != "" && util.NsxVersionLower(item.Metadata.IntroducedInVersion) {
```

Comment:
I think we need debug logging here and in the above conditional

Reply:
added